### PR TITLE
Slightly modify the detection of the number of CPU cores

### PR DESCRIPTION
**Pull request by rschu1ze**
AMD offers 48/67/64 core systems ([Milan](https://en.wikipedia.org/wiki/Epyc)), and soon CPUs with up to 96 cores (Genoa). Intel also sells [Xeon CPUs](https://www.intel.com/content/www/us/en/products/details/processors/xeon/scalable/platinum/products.html) with >32 cores. If the system has >= 32 logical cores then the current logic is to use only half of them. If SMT (HyperThreading) is on, then ClickHouse will effectively utilize only the physical cores (i.e. "disable" HyperThreading by means of the software).

However, if SMT is disabled (logical == physical core count), which is not uncommon, then ClickHouse uses only half the available physical cores. 

With this PR, the 32-core threshold is retained but the number of used cores is only reduced if SMT is on.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Remove the limitation that on systems with >=32 cores and SMT disabled ClickHouse uses only half of the cores

### CODE REVIEW

**Comment by rschu1ze** [comment]
@nickitat Do you know if SMP/hyperthreading generally benefits/deteriorates performance? In case of the latter, it will be better to check programmatically if SMP is on and halve the max used cores only in this case.

**Comment by nickitat** [comment]
looks like generally (e.g. if we take the whole set of clickbench queries) we should benefit from hyper threading. it was the case when we brought it back for 16 cores instances. it is because normal queries do have stalls on memory access and branch mispredictions. only heavy number crunching queries without lots of memory accesses (like Q29 from CB) could probably slow down, but it is fairly marginal case.
so if we run CB on machines with different number of cores I would bet that we will see better relative time with HT for both intel and amd even for `cpu_count >= 32`.

**Comment by rschu1ze** [comment]
Tnx, this makes sense.
Even if HT isn't beneficial, overhead will be low I guess (1-3%?).
(I could actually not find not much material how HT impacts other DBs, so I guess it is not a big point of contention.)

@alexey-milovidov Feel free to submit or advice further.

**Comment by rschu1ze** [comment]
Some funny measurements of ClickBench on a 128-vCore / 64 physical core instance (m6i.32xlarge):
- current HEAD: 64 threads --> user time: 3.463s
- not artificially reducing the thread number: 128 threads --> user time: 3.677s

(averaging 6 measurements each)

So with HyperThreadin,g CPU consumption is surprisingly 6% worse. Of course, this could be because ClickBench is CPU-intensive.

I nevertheless think that the code should be made aware of HT so we don't underutilize >32-core systems with HT disabled. We can keep the existing threshold though.

**Comment by alexey-milovidov** [comment]
Hyperthreading cores don't increase performance linearly.

If a query has a low number of cache misses and branch mispredictions, hyperthreading makes performance worse.
If a query has a high number of cache misses and branch mispredictions, hyperthreading makes performance better, but usually in the 10..50% range. Therefore, using all HT cores for a single query by default will worsen performance on many concurrent queries.

Cache misses mainly appear while doing large GROUP BY, DISTINCT, IN, and JOIN. But on a very large number of cores, the memory becomes the bottleneck earlier than we start using all hyper-threaded cores.

It will make sense to allow more cores by default, but queries should automatically stop using so many threads when concurrent queries start running. I can be done with https://github.com/ClickHouse/ClickHouse/pull/37285, but this setting is not enabled by default.

File: src/Common/getNumberOfPhysicalCPUCores.cpp [hunk 1]
```diff
@@ -36,20 +42,69 @@ static uint32_t getCGroupLimitedCPUCores(unsigned default_cpu_count)
...
+/// (HyperThreading) enabled, physical_concurrency() returns half of of std::thread::hardware_concurrency(), otherwise return the same.
+#if defined(__x86_64__) && defined(OS_LINUX)
+unsigned physical_concurrency()
+try
+{
+    /// The CPUID instruction isn't reliable across different vendors and CPU models. The best option to get the physical core count is
+    /// to parse /proc/cpuinfo. boost::thread::physical_concurrency() does the same, so use their implementation.
+    ///
+    /// See https://doc.callmematthi.eu/static/webArticles/Understanding%20Linux%20_proc_cpuinfo.pdf
+    std::ifstream proc_cpuinfo("/proc/cpuinfo");
```
Comment (alexey-milovidov) [inline src/Common/getNumberOfPhysicalCPUCores.cpp:55]: /proc is not necessarily mounted (chroot and FreeBSD).

File: src/Common/getNumberOfPhysicalCPUCores.cpp [hunk 2]
```diff
@@ -36,20 +42,73 @@ static uint32_t getCGroupLimitedCPUCores(unsigned default_cpu_count)
 }
 #endif
 
-static unsigned getNumberOfPhysicalCPUCoresImpl()
+/// Returns number of physical cores, unlike std::thread::hardware_concurrency() which returns the logical core count. With 2-way SMT
+/// (HyperThreading) enabled, physical_concurrency() returns half of of std::thread::hardware_concurrency(), otherwise return the same.
+#if defined(__x86_64__) && defined(OS_LINUX)
+unsigned physical_concurrency()
+try
```
Comment (serxa) [inline src/Common/getNumberOfPhysicalCPUCores.cpp:49]: Function-try-block looks strange to me w/o a constructor, but I guess it's just stylistic nuances.

File: src/Common/getNumberOfPhysicalCPUCores.cpp [hunk 3]
```diff
@@ -36,20 +42,73 @@ static uint32_t getCGroupLimitedCPUCores(unsigned default_cpu_count)
...
-#if defined(__x86_64__)
-    /// Let's limit ourself to the number of physical cores.
-    /// But if the number of logical cores is small - maybe it is a small machine
-    /// or very limited cloud instance and it is reasonable to use all the cores.
+#if defined(__x86_64__) && defined(OS_LINUX)
+    /// On really big machines, SMT is detrimental to performance (+ ~5% overhead in ClickBench). On such machines, we limit ourself to the physical cores.
+    /// Few cores indicate it is a small machine, runs in a VM or is a limited cloud instance --> it is reasonable to use all the cores.
     if (cpu_count >= 32)
-        cpu_count /= 2;
+        cpu_count = physical_concurrency();
```
Comment (serxa) [inline src/Common/getNumberOfPhysicalCPUCores.cpp:111]: This looks heavy. Let's use a static variable to parse the /proc/cpuinfo once?

File: src/Common/getNumberOfPhysicalCPUCores.cpp (hunk 2 above)
Comment (rschu1ze) [reply src/Common/getNumberOfPhysicalCPUCores.cpp:49]: It is a stylistic nuance :)

File: src/Common/getNumberOfPhysicalCPUCores.cpp (hunk 3 above)
Comment (rschu1ze) [reply src/Common/getNumberOfPhysicalCPUCores.cpp:111]: Good idea, it is implemented already (l. 126).

**Review by serxa — APPROVED** [review]

**Pull request merged**